src/main/java/net/coderbot/iris/pipeline/TransformPatcher.java

```diff
@@ -0,0 +1,512 @@
+						ParseTreeMatch match = entityColorPattern.match(ctx);
+						if (match.succeeded()) {
+							removeNode(ctx);
+						}
+					}
+				});
+
+				if (getJobParameters().type == ShaderType.GEOMETRY) {
+					chainDependent(new SearchTerminals<Parameters>().singleTarget(
+							new ParsedReplaceTargetImpl<>("entityColor", "entityColor[0]", GLSLParser::expression)));
+				}
+
+				chainDependent(new RunPhase<Parameters>() {
+					@Override
+					protected void run(TranslationUnitContext ctx) {
+						switch (getJobParameters().type) {
+							case VERTEX:
+								injectExternalDeclarations(InjectionPoint.BEFORE_DECLARATIONS,
+										"uniform sampler2D iris_overlay;",
+										"varying vec4 entityColor;");
+								break;
+							case GEOMETRY:
+								injectExternalDeclarations(InjectionPoint.BEFORE_DECLARATIONS,
+										"out vec4 entityColorGS;",
+										"in vec4 entityColor[];");
+								break;
+							case FRAGMENT:
+								injectExternalDeclaration(InjectionPoint.BEFORE_DECLARATIONS, "varying vec4 entityColor;");
+								break;
+						}
+					}
+				});
+			}
+		};
+
+		LifecycleUser<Parameters> wrapOverlay = new MainWrapper<Parameters>() {
+			@Override
+			protected String getMainContent() {
+				return getJobParameters().type == ShaderType.VERTEX
+						? "vec4 overlayColor = texture2D(iris_overlay, (gl_TextureMatrix[2] * gl_MultiTexCoord2).xy);\n" +
+								"entityColor = vec4(overlayColor.rgb, 1.0 - overlayColor.a);\nirisMain();"
+						: "entityColorGS = entityColor[0];\nirisMain();";
+			}
+		};
+
+		LifecycleUser<Parameters> renameEntityColorFragment = new SearchTerminals<Parameters>() {
+			@Override
+			public boolean isActive() {
+				return ((AttributeParameters) getJobParameters()).hasGeometry;
+			}
+		}.singleTarget(new TerminalReplaceTargetImpl<>("entityColor", "entityColorGS"));
+
+		// #endregion patchAttributes
+
+		// #region patchSodiumTerrain
+		// see SodiumTerrainPipeline for the original patcher
+		LifecycleUser<Parameters> wrapFTransform = new WrapIdentifier<Parameters>()
+				.wrapTarget("fTransform")
+				.detectionResult("iris_ftransform")
+				.injectionLocation(InjectionPoint.BEFORE_FUNCTIONS)
+				.injectionExternalDeclaration("vec4 iris_ftransform() { return gl_ModelViewProjectionMatrix * gl_Vertex; }");
+
+		LifecycleUser<Parameters> wrapVertex = new WrapIdentifier<Parameters>() {
+			@Override
+			protected ActivatableLifecycleUser<Parameters> getWrapResultDetector() {
+				return new SearchTerminals<Parameters>()
+						.targets(
+								Stream.of("a_Pos", "u_ModelScale", "d_ModelOffset")
+										.map(WrapThrowTargetImpl<Parameters>::new)
+										.collect(Collectors.toList()));
+			}
+
+			@Override
+			protected ActivatableLifecycleUser<Parameters> getInjector() {
+				return RunPhase.withInjectExternalDeclarations(injectionLocation(),
+						"attribute vec3 a_Pos;",
+						"uniform vec3 u_ModelScale;",
+						"attribute vec4 d_ModelOffset;");
+			}
+		}
```

Comment:
This might also be something we want to abstract a bit. I might also want to rename the `a_` / `u_` / `d_` to `iris_` on the side of the sodium compat code so we don't need special checks for it.

Reply:
Renaming to `iris_` everywhere would simplify things somewhat. Though with regular `WrapIdentifier` transformations, the checking is configured almost automatically.